### ISSUE TITLE
"circular require considered harmful" warning in Rollbar >= 2.19.3

### ISSUE THREAD

**Issue by jordan-brough** [rollbar/rollbar-gem#946]
As of Rollbar 2.20, Rollbar generates this warning:

> warning: loading in progress, circular require considered harmful - …/ruby/2.5.7/lib/ruby/gems/2.5.0/gems/rollbar-2.20.0/lib/rollbar/notifier.rb

Version 2.19.0:
```sh
# Rollbar 2.19.0
gem install rollbar -v 2.19.0
ruby -w -e 'gem("rollbar", "2.19.0"); require "rollbar"'
```
Does not produce a warning.

Version 2.20.0 (and any version since then):
```sh
# Rollbar 2.20.0
gem install rollbar -v 2.20.0
ruby -w -e 'gem("rollbar", "2.20.0"); require "rollbar"'
```
Produces this warning:
> …/ruby/2.5.7/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59: warning: loading in progress, circular require considered harmful - …/ruby/2.5.7/lib/ruby/gems/2.5.0/gems/rollbar-2.20.0/lib/rollbar/notifier.rb
> from -e:1:in  `&lt;main&gt;'
> …

I'm on Ruby 2.5.7p206.

It looks like https://github.com/rollbar/rollbar-gem/pull/773 fixed a similar issue back in 2018, after a report in https://github.com/rollbar/rollbar-gem/issues/544.

**Comment by waltjones** [comment]
@jordan-brough Thank you for the report. I'll take a look.

**Comment by jordan-brough** [comment]
@waltjones FYI I looked a bit more today and noticed that for both Ruby 2.5.7 and Ruby 2.6.5, the issue happens between `v2.19.2` and `v2.19.3`.

Specifically in commit 66ff7b9, where `Rollbar::Item::Locals` was added with `require 'rollbar/notifier` [here](https://github.com/rollbar/rollbar-gem/commit/66ff7b9fa69250e867d89540df46ab163751583d#diff-c74f5ee15c1b2e167f4ded70d4bc50faR1).

This makes the dependency chain go like this:

rollbar -> **notifier** -> item -> item/backtrace -> item/frame -> item/locals -> **notifier**

and Ruby complains when it hits the second `require 'rollbar/notifier'` in `rollbar/items/locals`.

**Comment by jordan-brough** [comment]
I also noticed that Ruby 2.7.0 has more "circular require considered harmful" complaints, even in `v2.19.2` https://gist.github.com/jordan-brough/2d7a550be3a52feb49e5fa0e412edafc. I haven't looked much at that, might be something to leave for another issue.

**Comment by waltjones** [comment]
@jordan-brough Thank you for the analysis.

**Comment by danroux** [comment]
I also experienced this in some of my projects, I investigated the issue and pushed a fix for it on: #1005. Hope it helps.

**Comment by waltjones** [comment]
Released in v3.1.2.
